### PR TITLE
BUG: parsing nanoseconds incorrect resolution

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

This is broken off from a branch that fixes a couple of higher-level bugs.

### CODE REVIEW

File: pandas/_libs/tslibs/parsing.pyx [hunk 1]
```diff
@@ -361,6 +371,33 @@ cdef parse_datetime_string_with_reso(
...
+    string_to_dts_failed = _string_to_dts(
+        date_string, &dts, &out_bestunit, &out_local,
+        &out_tzoffset, False
+    )
+    if not string_to_dts_failed:
+        if dts.ps != 0 or out_local:
+            # TODO: the not-out_local case we could do without Timestamp;
+            #  avoid circular import
+            from pandas import Timestamp
+            parsed = Timestamp(date_string)
```
Comment (mroeschke) [inline pandas/_libs/tslibs/parsing.pyx:385]: So the `dts` components cannot be passed into `Timestamp` and `date_string` needs to be passed instead to go through the string parsing flow?

File: pandas/_libs/tslibs/period.pyx [hunk 2]
```diff
@@ -2590,10 +2590,13 @@ class Period(_Period):
             dt = value
             if freq is None:
                 raise ValueError('Must supply freq for datetime value')
+            if isinstance(dt, Timestamp):
+                nanosecond = dt.nanosecond
         elif util.is_datetime64_object(value):
             dt = Timestamp(value)
             if freq is None:
                 raise ValueError('Must supply freq for datetime value')
+            nanosecond = dt.nanosecond
```
Comment (mroeschke) [inline pandas/_libs/tslibs/period.pyx:2599]: Do we have a test with `np.datetime64` objects passed? (Do they even support nanosecond resolution?)

Comment (jbrockmendel) [reply pandas/_libs/tslibs/period.pyx:2599]: yes and yes

File: pandas/_libs/tslibs/parsing.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/parsing.pyx:385]: i punted on trying to optimize the Timestamp construction (i.e. avoid re-parsing) bc i think this function needs a non-trivial refactor which is out of scope

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
hmm lots of failures.....

**Comment by jbrockmendel** [comment]
> hmm lots of failures.....

messed up fixing a merge conflict, now fixed+green

**Review by mroeschke — APPROVED** [review]

**Pull request merged**